docs/source/more_types.rst

```diff
@@ -933,3 +933,210 @@ and non-required keys, such as ``Movie`` above, will only be compatible with
 another TypedDict if all required keys in the other TypedDict are required keys in the
 first TypedDict, and all non-required keys of the other TypedDict are also non-required keys
 in the first TypedDict.
+
+Final attributes of classes and modules
+***************************************
+
+There are several situations where static guarantees about non-redefinition
+of certain names (or references in general) can be useful. One such example
+is module or class level constants, user might want to guard them against
+unintentional modifications:
+
+.. code-block:: python
+
+   RATE = 3000
+
+   class Base:
+       DEFAULT_ID = 0
+
+   # 1000 lines later
+
+   class Derived(Base):
+       DEFAULT_ID = 1  # this may be unintentional
+
+   RATE = 300  # this too
+
+Another example is where a user might want to protect certain instance
+attributes from overriding in a subclass:
+
+.. code-block:: python
+
+   import uuid
+
+   class Snowflake:
+       """An absolutely unique object in the database"""
+       def __init__(self) -> None:
+           self.id = uuid.uuid4()
+
+   # 1000 lines later
+
+   class User(Snowflake):
+       id = uuid.uuid4()  # This has valid type, but the meaning
+                          # may be wrong
+
+Some other use cases might be solved by using ``@property``, but note that both
+above use cases can't be solved this way. For such situations, one might want
+to use ``typing.Final``.
+
+Definition syntax
+-----------------
+
+The ``typing.Final`` type qualifier indicates that a given name or attribute
+should never be re-assigned, re-defined, nor overridden. It can be used in
+one of these forms:
+
+* The simplest one is ``ID: Final = 1``. Note that unlike gor generic classes
+  this is *not* the same as ``Final[Any]``. Here mypy will infer type ``int``.
+
+* An explicit type ``ID: Final[float] = 1`` can be used as in any
+  normal assignment.
+
+* In stub files one can omit the right hand side and just write
+  ``ID: Final[float]``.
+
+* Finally, one can define ``self.id: Final = 1`` (also with a type argument),
+  but this is allowed *only* in ``__init__`` methods.
+
+Definition rules
+----------------
+
+The are two rules that should be always followed when defining a final name:
+
+* There can be *at most one* final declaration per module or class for
+  a given attribute:
+
+  .. code-block:: python
+
+     ID: Final = 1
```

Comment:
It would be better to explicitly import `Final` at least in the first example that uses it.